Fixes #20057 - fix duplicate akey repo sets

jlsherrill opened the pull request:
this fixes duplicate repository sets from being returned
when listing them for an activation key.  In addition this improves performance
by using a single api call for all product content in the org,
as well as using a presenter to only fetch the activation key overrides
once, instead of once per product content.

In addition this updates the redhat repos page to use this new
content fetching method to speed up that page as well.

parthaa commented on app/models/katello/glue/candlepin/owner.rb:
```diff
@@ -38,6 +38,32 @@ def content_access_mode
       end
 
+      def enabled_product_content(for_repositories)
+        return [] if for_repositories && for_repositories.empty?
```
for_repositories.blank?

parthaa commented on app/models/katello/glue/candlepin/owner.rb:
```diff
@@ -38,6 +38,32 @@ def content_access_mode
+      def enabled_product_content(for_repositories)
+        return [] if for_repositories && for_repositories.empty?
+        repositories = for_repositories.select([:product_id, :content_id]).to_a
+        content_ids = repositories.map(&:content_id)
```
If you are just using content_id from here then why do you need [:product_id, :content_id]

parthaa commented on app/models/katello/glue/candlepin/owner.rb:
```diff
@@ -38,6 +38,32 @@ def content_access_mode
+      end
+
+      def filtered_product_content
+        cp_products = Katello::Resources::Candlepin::Product.all(Organization.first.label, [:id, :productContent])
```
I dont remember exactly how this works. Can you enlighten me on why you are getting all products from Organization.first  as opposed to current org?

jlsherrill replied:
that should be self ;)   /me thinks i copy and pasted from my console

jlsherrill commented:
@parthaa updated

parthaa commented on app/lib/katello/resources/candlepin.rb:
```diff
@@ -399,6 +399,10 @@ def import(organization_name, path_to_file, options)
           end
 
+          def product_content(organization_name)
+            Product.all(organization_name, [:id, :productContent])
```
This is no chance this would be confused with '::Katello::Product' right ?

jlsherrill replied:
I don't think so, we do things in similar places in this file

parthaa commented:
Code looks good to me. But the "show all " in the repository sets page seems to show me non rpm repos also (like I saw paths pointing to ostree and docker)

parthaa commented on app/models/katello/activation_key.rb:
```diff
@@ -113,16 +113,11 @@ def available_content(content_access_mode_all = false, content_access_mode_env =
-      end
-      content.uniq
+      considered_products = content_access_mode_all ? all_products : self.products
+      repositories = Katello::Repository.where(:product_id => considered_products)
```
Might need to limit this to type yum.

jlsherrill replied:
fixed by using .subscribable

parthaa commented:
I definitely notice the perf improvements definitely and no dupes. Nice work. Address the above comment and should be good to go.

parthaa commented on app/models/katello/glue/candlepin/owner.rb:
```diff
@@ -38,6 +38,31 @@ def content_access_mode
+        to_return = []
+
+        cp_products.each do |product_hash|
+          product = ::Katello::Product.find_by(:cp_id => product_hash['id'])
```
cp_id is not org safe!

parthaa replied:
```
-          product = ::Katello::Product.find_by(:cp_id => product_hash['id'])
+          product = ::Katello::Product.find_by(:organization_id => self.id, :cp_id => product_hash['id'])
```

parthaa commented:
Not to overcomplicate this PR but can you check if the content hosts could use this logic for subs and repo sets... My guess is they are suffering from the same ailments

parthaa commented:
Rh Repos page
```
Before PR
 Completed 200 OK in 26777ms (Views: 26283.9ms | ActiveRecord: 85.3ms)

After PR
 Completed 200 OK in 8058ms (Views: 6534.3ms | ActiveRecord: 1156.0ms)
```

jlsherrill commented:
Updated to fix multi-org issue, added to content host repo sets listing, and sped it up even more by using more specific includes on the candlepin request.

jlsherrill commented:
Updated to move to a service as much as i could.  let me know if you like it better

parthaa commented on app/controllers/katello/api/v2/activation_keys_controller.rb:
```diff
@@ -136,7 +136,18 @@ def available_releases
+      content = pcf.product_content
+
+      overrides = @activation_key.content_overrides
+      content.map! { |pc| ProductContentPresenter.new(pc, overrides) }
```
I like the refactoring but I think you could DRY it up further by adding overrides logic to the Finder and make him return ProductContentPresenter... That would mean we can use the same logic for both AK and CH

jlsherrill commented:
@parthaa updated!

parthaa commented:
APJ
